### DISCUSSION TITLE
Best way to add a virtual IP for use in a VPN tunnel?

### DISCUSSION THREAD

**Discussion by misch42** [strongswan/strongswan#1376] [General]
Hi,

I have a special problem. I want to connect one single host to a network. Should be simple. BUT:

The other side asked my to use a RFC1918 address as a source address in the tunnel.

So I have an official IP address on the network to the default gateway. And I want to have an additional IP address on that host for use within the tunnel.

At the moment I am doing the following:
1) I add the tunnel address to the lo interface.
2) I add a route to use this address as source if routing to the partner network:
ip r a <other.net.address> via < gateway > dev <interface> src <rfc1918.address> onlink

Is these a more simple solution o achieve this?

Michael

**Comment by tobiasbrunner** [comment]
Sounds about right. Just install the IP on a local interface, then use it in the config as local traffic selector. You don't have to install a route manually if the negotiated IPsec policies are matching, as strongSwan will install one in table 220 (check with `ip route list table 220`).

**Reply by misch42** [reply]
Strange. setup local_ts and remote_ts. Child SAS and policies are installed. But routing table 220 is empty:

# ip a l
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 brd 127.255.255.255 scope host lo
       valid_lft forever preferred_lft forever
    inet <internal-IP>/32 scope global lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: enp41s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether <mac> brd ff:ff:ff:ff:ff:ff
    inet<external-IP>/32 scope global enp41s0
       valid_lft forever preferred_lft forever
    inet6 fe80::aaa1:59ff:fe94:f1c1/64 scope link 
       valid_lft forever preferred_lft forever

# ip r l
default via <gateway> dev enp41s0 proto static onlink 
<remote-net>/26 via <gateway> dev enp41s0 proto static src <internal-IP> onlink <---- route added manually in config


# ip xfrm policy
src <internal-IP>/32 dst <remote-net>/26 
        dir out priority 370303 
        tmpl src <external-IP> dst <remote-IP>
                proto esp spi 0x2217f419 reqid 3 mode tunnel

# ip route list table 220
<empty>

Ideas?

**Reply by tobiasbrunner** [reply]
Read the log.

**Reply by Thermi** [reply]
> /26 via dev enp41s0 proto static src onlink <---- route added manually in config

That breaks it. Don't do that.

strongSwan already puts in the right route into table 220

**Reply by misch42** [reply]
without this strongswan complains about not beeing able to add the route during child setup:
[KNL] received netlink error: Network is unreachable (101)
[KNL] unable to install source route for <local-net>

This is because the server is net up somehow, well, unconventional. The IP address is in a /32 network. I suppose this is due to some restrictions by the cloud hoster:

System networking:

# ip a l
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen
(...)
2: enp41s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether <MAC> brd ff:ff:ff:ff:ff:ff
    inet <extrenal-IP>/32 scope global enp41s0   <---- here is the problem!
       valid_lft forever preferred_lft forever
    inet6 fe80::aaa1:59ff:fe94:XXXX/64 scope link 
       valid_lft forever preferred_lft forever
# ip r l
default via <gateway> dev enp41s0 proto static onlink    <---- default routing also via "onlink" option.

Already the system network setup has to utilize the "onlink" option to make the default routing working.

This server is set up at a major German hosting company. I suppose it is some virtual machine. 
When I add the VPN source route in the _updown script, it seems to work. At least I see some encrypted packed leaving the external interface. I have to setup my own remote server to check, if the correct packets are send to the remote side.

**Reply by tobiasbrunner** [reply]
Yes, as mentioned below, such `onlink` routes are problematic. In your case, is there really no address after the `via` in the route? (I noticed you just omitted other stuff, so not sure.) If not, and strongSwan uses the address as next hop in its own route (increase the log level for _knl_ to see that), you could add a non-`onlink` route for that address/subnet (something like `<nexthop or its subnet> dev enp41s0 proto kernel scope link src <ip on device>`), which should allow strongSwan to install its routes.

**Reply by misch42** [reply]
Hi,

I checked it with a remote server that I do control. Everything works as designed. I have to ass the route manually (or in the updown script) including the onlink option. On the other side I see packets with the internal IP source address. And yes, I have to include the gateway in the route command. I just removed it due to privacy.

**Comment by misch42** [comment]
ok. since this machine has such as stupid set the default route is added with the onlink option. of, course, strongswan cannot catch this also fails. Perhaps the best option would be to add the route in table 220 in the _updown script Will try.

**Reply by tobiasbrunner** [reply]
The `onlink` default route could definitely be a problem (e.g. result in a `Network is unreachable` error from the kernel when the source route is installed without that flag, because there is no regular route to the next hop). You can check the log (if there is an error it should already be logged with the default levels, for more details, use the recommended log levels, see #196).

**Reply by Thermi** [reply]
It most definitely is the cause of the problems here

**Comment by misch42** [comment]
See my previous comment. I think I am able to add the route in the _updown scipt. Perhaps the project could offer a config option in swanctl.conf to use the "onlink" option in these cases.

I will have to check if everything works when I add the VPN route manually with a remote gateway that I control. I will keep you updated.
